### ISSUE TITLE
macOS event dispatching broken

### ISSUE THREAD

**Issue by TheJJ** [SFTtech/openage#852]
#828 figured out that for a successful mac run the event dispatching for mac needs to be changed.
This has to be fixed properly.

**Comment by Birch-san** [comment]
Useful references…

- The [magic commit](https://github.com/ChipmunkV/openage/commit/d577f5681bf61e09cd7d8175661be2df389b9b4d).

- https://github.com/SFTtech/openage/pull/828#issuecomment-299688716:

> that magic partially disables input handling, so it needs some work: https://github.com/SFTtech/openage/issues/819#issuecomment-298831187

- https://github.com/SFTtech/openage/issues/819#issuecomment-298831187:

> [Birch-san] @ChipmunkV, I know you've said that the commit was removed because it was unexplainable. But where did you get the idea to do that? Did you read something somewhere?
>
> [ChipmunkV] @Birch-san, I knew that event loops in Qt use event dispatchers that have OS-specific implementation. Then you just disable the input handling part and try.
>
> I think that the issue is easily fixable by implementing and installing a custom dispatcher with [`QCoreApplication::setEventDispatcher()`](http://doc.qt.io/qt-5/qcoreapplication.html#setEventDispatcher).

**Comment by Birch-san** [comment]
I actually had a stab at ChipmunkV's idea (installing a custom dispatcher):

https://github.com/SFTtech/openage/compare/master...Birch-san:set-event-dispatcher

I have no idea what I'm doing. At the time of writing: my custom event dispatcher compiles, and openage runs, but we still have a white screen followed by a black screen.  
I also haven't checked whether my event dispatcher is installed correctly. It needs to be installed _before_ you init the QCoreApplication. I _assume_ I've done that, but haven't actually verified.

A good reference for how to build a custom event dispatcher is here (MIT License, peper0@GitHub):

https://github.com/peper0/qtasio/blob/master/qasioeventdispatcher.cpp
https://github.com/peper0/qtasio/blob/master/qasioeventdispatcher.h

Note: peper0's code includes private headers such as:

```cpp
#include "private/qabstracteventdispatcher_p.h"
#include "private/qtimerinfo_unix_p.h"
```

We're not supposed to do that. It has the following message:

> This file is not part of the Qt API.  It exists purely as an
> implementation detail.  This header file may change from version to
> version without notice, or even be removed.
> 
> We mean it.

I think we can live without those private headers anyway. [`qabstracteventdispatcher_p.h`](https://code.woboq.org/qt5/qtbase/src/corelib/kernel/qabstracteventdispatcher_p.h.html) is basically empty. [`qtimerinfo_unix_p.h`](https://code.woboq.org/qt5/qtbase/src/corelib/kernel/qtimerinfo_unix_p.h.html) is more substantial, but whatever.

**Comment by Birch-san** [comment]
I also considered "maybe a custom event dispatcher is overkill. could we just filter out events instead?"

https://github.com/SFTtech/openage/compare/master...Birch-san:event-flags

I believe no matter what combination of events that I filtered out: I still ended up with a black screen, and the animated graph never appeared. These are the notes that I took at the time:

```
WARN QFlags(0x1|0x2|0x4) = no graph
WARN QFlags(0x2|0x4) = no graph
WARN QFlags(0x4) = no graph
WARN QFlags(0x1|0x2) = no graph
WARN QFlags(0x1) = no graph
WARN QFlags(0x2) = no graph
WARN QFlags(0x1|0x4) =
```

Maybe I should've also tried filters on `gui_event_queue_impl.cpp`, rather than just on `gui_application_impl.cpp`.

**Comment by Birch-san** [comment]
There's also this:

- [`QCoreApplication::installNativeEventFilter`](https://doc.qt.io/qt-5/qcoreapplication.html#installNativeEventFilter)
- [`QAbstractNativeEventFilter::nativeEventFilter`](https://doc.qt.io/qt-5/qabstractnativeeventfilter.html#nativeEventFilter)

I wonder if the event dispatch problem is just with some platform-specific events. So maybe we could see some change if we filtered them out.

**Comment by Birch-san** [comment]
Btw, the magic commit is no panacea. This is a video of what openage looks like with the magic commit:

https://www.youtube.com/watch?v=9GmwKreWg6I

Ignore the blue tint — that's an endianness problem that is now fixed.  
Rather, notice that loads of interface elements disappear/reappear all the time.

https://github.com/SFTtech/openage/pull/756#issuecomment-292804797:

> [ChipmunkV]: A theory about the disappearing buttons: it's probably related to that input handling fix - some events can't pass through if there is not enough real input activity (so the GUI isn't updated).

**Comment by ChipmunkV** [comment]
Btw, I've started the event dispatcher too. And probably I've started from that piece of code where it's created (to make sure that it's created before the QApplication).

The `gui_dedicated_thread.cpp` that you're changing is a dead code for the future case where the GUI event loop is separated from the rendering thread (sorry about that, it's not a good practice to leave stuff "for the future" on the main branch).

The possible places to attach a dispatcher are: `libopenage/gui/integration/public/gui_application_with_logger.cpp` (the logger is also attached before the QApplication is created) or `libopenage/gui/guisys/private/gui_application_impl.cpp`.

About the events: the OS-level event loop is handled by SDL, input events for Qt are synthesized by the openage from the SDL events. There shouldn't be any way for the events from OS to get into Qt.

**Comment by Birch-san** [comment]
I continued to investigate this.

I've put useful stuff into this branch, you can comment out bits that you don't want to use:

https://github.com/SFTtech/openage/compare/master...Birch-san:event-dispatch-2

The main things that you may find yourself commenting out are:

- `gui_application_impl.cpp`: installation of the custom event dispatcher (because it's no-op)
  - remove `app{(QCoreApplication::setEventDispatcher(&this->event_dispatcher), argc), &argv}`
  - prefer `app{argc, &argv}`
- the logging that is performed in my event filters (because they're noisy; depends if that's what you're testing)
  - remove `KeyPressEater.cpp` `qInfo() << "EVENT: " << event->type();`
  - remove `CocoaNativeEventFilter.cpp` `qInfo() << "NATIVE EVENT: " << eventType << "; " << [event type];`
- `gui_application_impl.cpp`: installation of the native event filter (if you're testing this on non-macOS for some reason, and can't compile Objective-C):
  - remove `this->app.installNativeEventFilter(&this->native_event_filter);`

I've demonstrated here:

- how to install an event filter
  - currently I just use this to log events
- how to install a native macOS event filter
  - currently I just use this to log native events
- what a no-op event dispatcher looks like
  - installed using @ChipmunkV's technique

Things that I noticed…

- When GuiApplicationImpl invokes `QGuiApplication#processEvents()`: it never finishes.
  - well, maybe this is what it's supposed to do?
- When GuiEventQueueImpl invokes `QEventLoop#processEvents()`: it never finishes.
  - well, maybe this is what it's supposed to do?
- I tried setting a 100ms maxtime on processEvents: `this->app.processEvents(QEventLoop::AllEvents, 100);`
 - this did _not_ help (the function call _still_ never finished).
- Similarly: I tried toggling various event flags (as I mentioned in an earlier comment), one of which is`QEventLoop::WaitForMoreEvents`. No combination that I tried did anything. I tried on _both_ `processEvents()` invocations.
- If _either_ of those classes invokes their `processEvents()`: we will never again see any events logged in our event filter.
  - probably this means we never process any QTimer events. I assume this is why the graphics don't update.
  - we _do_ continue to see native events (mouse move, keypresses) logged in our native event filter
  - maybe I'm interpreting this incorrectly though. perhaps `processEvents()` isn't meant to pass events through the filter first?
- I tried filtering out all events and all native events. This did not solve any problems.
  - So this means there's no "bad event" that the event dispatcher fails to handle.
- `QThread::currentThread() == QCoreApplication::instance()->thread()` is _always_ true, as far as I noticed
  - this logic was used by the "magic commit" to ensure that `processEvents()` was never run on the QCoreApplication's thread
  - in practice: I think it just means that `processEvents()` is never run, _ever_.
  - the reason this helps is because `processEvents()` seems to be broken in some way with the default macOS event dispatcher.

In case you're interested, this is what I observe when I use the no-op event dispatcher:
- If `processEvents()` is run: we see neither events, nor native events logged in our event filter
- If `processEvents()` is not run: 
  - we see both events and native events logged in our event filter
  - we see the real-time graph, but nothing else:

<img width="1552" alt="image" src="https://user-images.githubusercontent.com/6141784/28499258-77e5a7bc-6fa9-11e7-83ad-90b2442e4743.png">

So, that's various things that didn't work.

I do caution that writing a custom event dispatcher could be _very_ hard. [Here's](https://gist.github.com/Birch-san/7cebf4863873bd1039f9c6e00eb1873c) `qt5/qtbase/src/corelib/kernel/qeventdispatcher_cf.mm`, which I assume is Qt's default event dispatcher implementation for macOS (i.e. it's the only one that I see written in Objective-C).

I see that Qt5 has the following event dispatcher implementations:

```
qt5/qtbase/src/corelib/kernel/qeventdispatcher_cf_p.h
qt5/qtbase/src/corelib/kernel/qeventdispatcher_cf.mm
qt5/qtbase/src/corelib/kernel/qeventdispatcher_glib_p.h
qt5/qtbase/src/corelib/kernel/qeventdispatcher_glib.cpp
qt5/qtbase/src/corelib/kernel/qeventdispatcher_unix_p.h
qt5/qtbase/src/corelib/kernel/qeventdispatcher_unix.cpp
qt5/qtbase/src/corelib/kernel/qeventdispatcher_win_p.h
qt5/qtbase/src/corelib/kernel/qeventdispatcher_win.cpp
qt5/qtbase/src/corelib/kernel/qeventdispatcher_winrt_p.h
qt5/qtbase/src/corelib/kernel/qeventdispatcher_winrt.cpp
```

Maybe we could use the UNIX one or the glib one somehow. But I assume the macOS one exists for a reason.

Some of this stuff is private, so it may be non-trivial to copy-paste. Maybe we can construct an instance and delegate to it?

You can have a look at that source like so:

```bash
git clone git://code.qt.io/qt/qt5.git
cd qt5
git checkout v5.8.0
cd qtbase
git submodule update --init --recursive .
```

But seriously I've no idea what I'm doing.

**Comment by Birch-san** [comment]
https://doc.qt.io/qt-5/qeventloop.html#exec

> More sophisticated idle processing schemes can be achieved using processEvents().

Do we definitely need to use `QEventLoop::processEvents()`? could we switch to something like `QEventLoop::exec()`? sounds like it's less complex. if it meets our requirements: perhaps using it would randomize the situation enough that the problem disappears.

**Comment by ChipmunkV** [comment]
That's where I'm at: https://github.com/ChipmunkV/openage/commit/13ad7630180539b06ee24e22d080933046d025df

The most relevant part is `libopenage/gui/guisys/private/gui_timer_info.cpp`. Need to make it work so the dispatcher can process timers. That will probably be enough.

`QGuiApplication::processEvents`, `QEventLoop::processEvents` and `QEventLoop::exec` call our `QEventDispatcherImpl::processEvents` on Linux for me. I'll try your code.

**Comment by Birch-san** [comment]
There's not really much in my code. All it does is:

- install event filters (I'm only using this for logging)
- install a no-op event dispatcher

But it seems you're significantly past no-op. I'll checkout your commit and see whether it does anything interesting on macOS so far.

**Comment by Birch-san** [comment]
I encountered this at compile-time:

```
/Users/birch/git/openage/libopenage/gui/guisys/private/gui_timer_info.cpp:119:6: error: use of undeclared identifier 'QTimerInfoList';
      did you mean 'TimerInfoList'?
void QTimerInfoList::repairTimersIfNeeded()
     ^~~~~~~~~~~~~~
     TimerInfoList
```

Changed it to TimerInfoList, encountered [this SIGSEGV](https://gist.github.com/Birch-san/bb7f91100c11e9857d6c98fe77765d31) when launching the app. Maybe that's expected behaviour, but reporting it just in case macOS run information is useful at this stage.

**Comment by Birch-san** [comment]
regarding `gui_timer_info.cpp`: where does your commented-out implementation come from?

Is it from [Qt4.8's `qeventdispatcher_unix.cpp`](http://cep.xray.aps.anl.gov/software/qt4-x11-4.8.6-browser/d9/dbe/qeventdispatcher__unix_8cpp_source.html)?

**Comment by ChipmunkV** [comment]
That branch is not runnable, but I think that it should crash in some different way.

The dispatcher is not from the `qeventdispatcher_unix.cpp`. I've made some socket handling with `select()` because the application was requesting sockets notifiers.

Then it requested to add timers, so I've taken a `qtbase/src/corelib/kernel/qtimerinfo_unix.cpp` from the Qt5 implementation. Rewriting it to `chrono`.

I've launched your version, and it seems to do fine on Linux. The `processEvents` of the custom dispatcher gets called, GUI works (timers don't work). Also, there are two problems I see: the custom event dispatcher has to be created with `new`, the `int events;` in the event dispatcher should be initialized

With this version the`QGuiApplication::processEvents()`and `QEventLoop::processEvents()` never finishes on macOS?

P. S. I think that all the code for this change has to be platform-independent.

**Comment by Birch-san** [comment]
When I make the changes that you recommended (https://github.com/Birch-san/openage/compare/event-dispatch-2...Birch-san:event-dispatch-3):

- `QEventDispatcherImpl.cpp`: `int events = 0;`
- `gui_application_impl.cpp`: `app{(QCoreApplication::setEventDispatcher(new QEventDispatcherImpl), argc), &argv}`

I find that the game actually works exactly like it does with the magic commit. It even works if I remove these guards `if (QThread::currentThread() == QCoreApplication::instance()->thread()) return;` from `gui_application_impl.cpp` and from `gui_event_queue_impl.cpp`.

`processEvents()` actually finishes now, and runs in a continuous loop:

```
WARN Finished processing of local event loop for thread:  QThread(0x7f8b5e6a6ba0)
WARN Processing of GUI application events will continue for thread:  QThread(0x7f8b5e6a6ba0)
WARN processFlags()
WARN Finished processing of GUI application events for thread:  QThread(0x7f8b5e6a6ba0)
WARN Processing of event queue will continue for thread:  QThread(0x7f8b5e6a6ba0)
WARN processFlags()
```

So… perhaps the no-op event dispatcher is significantly better than I realised.

I guess there's still the remaining problem of "UI elements keep disappearing" (just like in the magic commit).

**Comment by ChipmunkV** [comment]
Ok, then it's `int events;` that was causing the loop to block.

If the UI elements still keep disappearing then there is probably a rendering bug with the FBO: it's given to the Engine too early. Whole scenegraph nodes are missing. There is a debug mode to see separate elements: `QSG_VISUALIZE=overdraw ./run` (http://doc.qt.io/qt-5/qtquick-visualcanvas-scenegraph-renderer.html#visualizing-overdraw)

Maybe `glFinish()` in `GuiRendererImpl::render()` in `libopenage/gui/guisys/private/gui_renderer_impl.cpp` will change something.

But we still can't rule out the processEvent-related cause because the UI elements aren't flickering every frame. They suddenly disappear after the event processing.

**Comment by Birch-san** [comment]
Wow, glFinish() did the trick!

**Comment by Birch-san** [comment]
This is where I put glFinish():

https://github.com/Birch-san/openage/compare/event-dispatch-3...Birch-san:event-dispatch-4

**Comment by Birch-san** [comment]
I did try the `QSG_VISUALIZE=overdraw` also.

Whenever any UI elements disappeared: the visualization disappeared also. I also some cases where more UI elements were visible in reality, than were visible in the visualization:

![image](https://user-images.githubusercontent.com/6141784/28503137-c46e1052-6ff8-11e7-8f24-545747af4abd.png)

**Comment by ChipmunkV** [comment]
Currently both GL contexts are in one thread. Linux drivers are okay if there is no GL synchronization used in this case. But macOS drivers are probably using some intermediate per-context GL command queues, so they need sync in this case too.

**Comment by Birch-san** [comment]
> But macOS drivers are probably using some intermediate per-context GL command queues, so they need sync in this case too.

Is that what the glFinish() does? Or is there some other problem remaining?

**Comment by ChipmunkV** [comment]
> Is that what the glFinish() does? Or is there some other problem remaining?

Yes, it does that. Just need to read Internet on how people use FBOs with `glFinish`.

**Comment by Birch-san** [comment]
Does the "poor man's event dispatcher" do everything we need it to do? it doesn't implement timers, it doesn't implement socket notifiers. the only thing it really does is call `QCoreApplication::sendPostedEvents()` until there are no events. also there's logic (`events`, `total_events`) I'm pretty sure isn't used.

also it looks like it uses glib:

```cpp
//QCoreApplicationPrivate::sendPostedEvents(0, 0, d->threadData); //unix dispatcher use this
QCoreApplication::sendPostedEvents(); //glib dispatcher call this after every call of "awake".
```

Not sure whether `sendPostedEvents()` is _definitely_ implemented using glib, or if they actually meant "default event dispatch mechanism that Qt provides for your platform".

Is it okay for this to use glib? is that fine for Windows? doesn't seem like we had to add any dependencies to make this work.

**Comment by Birch-san** [comment]
Maybe a less hacky solution (i.e. one that actually implements a full event dispatcher) would be to find a way to explicitly ask for the glib-based event dispatcher described in `qt5/qtbase/src/corelib/kernel/qeventdispatcher_glib_p.h`,`qeventdispatcher_glib.cpp`, rather than allowing Qt to give us the default macOS implementation, (`qeventdispatcher_cf.mm`, `qeventdispatcher_cf_p.h`). not sure how. but certainly there could be unforeseen problems with using a nearly-empty implementation.

**Comment by ChipmunkV** [comment]
`sendPostedEvents()` is delivering events posted with `QCoreApplication::postEvent()`. For me it was always a pure Qt thing. We're using `QCoreApplication::postEvent()` to manually pass keyboard/mouse input to Qt (in `gui_input_impl.cpp`).

There is no API for asking for the platform-specific dispatchers. I think that it's more robust to write a cross-platform implementation of the dispatcher and use it on all platforms.

The goal of the event dispatchers is to get input events from the platform, but we're passing SDL input events through the `QCoreApplication::postEvent()` hack, which means that our real platform is SDL. We can reimplement `gui_input_impl.cpp` as an event dispatcher or we can make a timers-sockets-only event dispatcher and leave `gui_input_impl.cpp` as is.

One last thing that bothers me: when we were doing this things without custom event dispatcher:
```
#ifndef __APPLE__
 	this->app.processEvents();
#endif
...
#ifdef __APPLE__
	if (QThread::currentThread() != QCoreApplication::instance()->thread())
		this->callback_processor.processEvents();
#else
...
```
so there are no calls to `processEvents()` anywhere for `__APPLE__`. What was processing events on macOS? If the GUI kinda worked that way then something was calling `QCoreApplication::sendPostedEvents()`?

**Comment by Birch-san** [comment]
I checked which DTrace probes were exposed for the `sendPostedEvents()` function:

```bash
sudo dtrace -p "$(pgrep run)" -ln 'pid$target:QtCore:*sendPostedEvents*:entry {}'
```
```
   ID   PROVIDER            MODULE                          FUNCTION NAME
 8037   pid53854            QtCore QCoreApplication::sendPostedEvents(QObject*, int) entry
 8038   pid53854            QtCore QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) entry
```

I captured each distinct stack (4 frames deep), and how many times that stack occurred:

```bash
sudo dtrace -p "$(pgrep run)" -n 'pid$target:QtCore:*sendPostedEvents*:entry { @[ustack(4)] = count(); }'
```

When we use the poor-man's event dispatcher, and allow all our calls to `processEvents()`, we get [this variety of stack traces](https://gist.github.com/Birch-san/9f5c1ef103bde4086a6055bc0a9cb2d0).

When we use the magic commit (i.e. where we do _not_ install a custom event dispatcher, but _do_ delete our call to `processEvents()` in `gui_event_queue_impl.cpp` and in `gui_application_impl.cpp`), we get [this variety of stack traces](https://gist.github.com/Birch-san/c8b3de1e117c49a7fd3167470d949e68).

And just for fun, here's what it looks like if we do **both** (i.e. where we install a custom event dispather **and** delete our calls to processEvents()): [stack traces](https://gist.github.com/Birch-san/3399a5c67a0d95b3c543d6888a972a04).

**Comment by ChipmunkV** [comment]
From looking at the results of the magic commit, we see that a call through the SDL2 (presumably `SDL_PollEvent`) gets to the `QCoreApplicationPrivate::sendPostedEvents`. Qt somehow sneaks a handler into cocoa. I haven't expected it to ever do that.

With event dispatcher and allowing calls to `processEvents()` it looks sane - all is handled by the QEventDispatcherImpl``. The other forgotten `QEventLoop`s that are in separate threads (the `RecursiveDirectoryWatcher` thread to watch for qml files modifications and some internal thread of `QtQml`) receives an `QEventDispatcherUNIX` that does the right thing for it.

So we can fix all by adding `glFinish()` and a custom event dispatcher for all platforms.

Need to just implement the timers in the dispatcher. Can test it by pasting http://doc.qt.io/qt-5/qml-qtqml-timer.html example somewhere in the openage qml code:
```
Item {
    Timer {
        interval: 500
        running: true
        repeat: true
        onTriggered: time.text = Date().toString()
    }

    Text { id: time }
}
```

**Comment by Birch-san** [comment]
I think we may not need to implement our own timer. I was able to integrate the default `QTimerInfoList`.

https://github.com/Birch-san/openage/compare/event-dispatch-4...Birch-san:event-dispatch-6

I tried putting that timer UI in a few places. First I tried sane places, but it wasn't visible (maybe default font color is black? tried changing that, but didn't help). Next I put it in progressively more horrifying places until it finally showed up:

<img width="1552" alt="image" src="https://user-images.githubusercontent.com/6141784/28757615-9dc59aca-757e-11e7-8712-549390b9f395.png">

The timer updates live, so I assume that confirms that the timer events work.

Sometimes (50% of the time?) no QML ever appears:

<img width="1552" alt="image" src="https://user-images.githubusercontent.com/6141784/28757611-880cf55c-757e-11e7-938c-81575df75b3c.png">

The first time I noticed that problem was as of commit https://github.com/Birch-san/openage/commit/3c081a8b39eb2f62720e9b28b21157e14781d119:

<img width="1000" alt="image" src="https://user-images.githubusercontent.com/6141784/28757625-cd1da9ca-757e-11e7-8045-896a5bc2cf53.png">

Basically I took the implementation of processEvents() that you provided (which implemented socket notifiers), and added timers by referring to `qeventdispatcher_unix.cpp`:

https://code.woboq.org/qt5/qtbase/src/corelib/kernel/qeventdispatcher_unix.cpp.html
https://code.woboq.org/qt5/qtbase/src/corelib/kernel/qeventdispatcher_unix_p.h.html

However there are parts of its logic which I skipped. Maybe that explains why the QML sometimes fails to appear.

If we take the approach from my branch: we would become dependent on qtimerinfo_unix_p.h:

https://code.woboq.org/qt5/qtbase/src/corelib/kernel/qtimerinfo_unix_p.h.html
https://code.woboq.org/qt5/qtbase/src/corelib/kernel/qtimerinfo_unix.cpp.html

Also it means incorporating some LGPL3 header code from Qt (maybe this is fine; we are GPL3, which I assume is a superset).

I don't see a version of QTimerInfo for any other platform. Is this why you recommended that we write our own cross-platform implementation? that seems like it would be a reasonably big job. especially since it concerns time: we've had experiences with time code failing to compile on macOS (or worse: failing to compile on slightly old versions of macOS). so a cross-platform timerinfo may be difficult.

Trying to figure out where to go from here. I assume that if I copy `qeventdispatcher_unix` a bit more closely: there's a chance that it'll fix the disappearing QML. and then we'd have a working event dispatcher for macOS (and for other UNIX). but I assume we'd then break Windows. and I don't immediately see anything we could copy to make Windows work. which suggests maybe the easier solution is to have platform-specific event dispatch (for example: just install this event dispatcher for macOS, or just for UNIX).

What are the repercussions of just using the magic commit + glFinish()? would it mean we simply have no event processing? the game seemed to work despite that.

**Comment by ChipmunkV** [comment]
> I don't see a version of QTimerInfo for any other platform. Is this why you recommended that we write our own cross-platform implementation?

Other platforms are probably using the OS-specific API to implement timers. The `qtimerinfo_unix` is a POSIX code that I understand. And I also understand how to rewrite it with C++ std::chrono (which is probably a requirement since, as you've mentioned, Windows is only partially POSIX). 

> What are the repercussions of just using the magic commit + glFinish()? would it mean we simply have no event processing?

I haven't tested, but I think that the GUI animations will play instantly or won't play at all. To check that, set the `interval`of that `Timer` in QML to something like 5000ms instead of 500 and see if it really updates every 5 seconds.

**Comment by TheJJ** [comment]
I think this was now fixed by #995, please test.
